### PR TITLE
Support splittable for CSV format in file connector [HZ-1405]

### LINKED ISSUE #21854: Row count is exactly 4 times higher from S3 than original row count in the file [HZ-1405]
**Describe the bug**
Row count is exactly 4 times higher when using in SQL file connector to read from S3 than original row count in the file

**Expected behavior**
Row count should always same for the same file.

**To Reproduce**

1. Upload a ~2.7 mil rows csv file into S3 bucket.
2. Create a file mapping locally to connect to the file in S3
3. Select row count from the mapping

**Additional context**
SQL to use
```
CREATE MAPPING ustest
TYPE File
OPTIONS (
    'path' = 's3a://my-bucket/',
    'sharedFileSystem' = 'true',
    'format' = 'csv',
    'glob' = 'ustest.csv',
    'fs.s3a.endpoint' = 's3.region.amazonaws.com',
    'fs.s3a.access.key' = 'abc',
    'fs.s3a.secret.key' = 'efg'
);
```
Screenshot 2022-07-27 195245
Screenshot 2022-07-27 195221

### PR DESCRIPTION
**Pull request by frant-hartm**
When the input file is large it is split into multiple FileSplits. Previously the CsvInputFormat wasn't respecting the split boundaries and read whole file. Causing duplicates when file had multiple splits.

Migrated to LineRecordReader, which handles splits and is suitable to read file line by line, required for CSV.

Additional logic was added to the `initialize` method to handle first line of the CSV file containing the header. It is used either to remap the order of fields or to build a schema.

Also added tests for non splittable (gzip) and splittable (bzip2) compression.

Fixes #21854

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible
- [ ] Send backports/forwardports if fix needs to be applied to past/future releases

Backports to 5.1 and 5.0

### CODE REVIEW

File: extensions/hadoop/src/main/java/com/hazelcast/jet/hadoop/impl/CsvInputFormat.java [hunk 1]
```diff
@@ -54,54 +59,87 @@ public RecordReader<NullWritable, Object> createRecordReader(InputSplit split, T
...
-            @Override
-            public boolean nextKeyValue() {
-                if (!iterator.hasNext()) {
-                    return false;
+            private String[] readHeader(FileSplit fileSplit, TaskAttemptContext context) throws IOException {
+                Configuration configuration = context.getConfiguration();
+                Path file = fileSplit.getPath();
+                FileSystem fs = file.getFileSystem(configuration);
+
+                try (InputStream in = fs.open(file)) {
```
Comment (TomaszGaweda) [inline extensions/hadoop/src/main/java/com/hazelcast/jet/hadoop/impl/CsvInputFormat.java:113]: you can merge those try blocks:
```java
ObjectReader headerReader = new CsvMapper()
                                .enable(Feature.WRAP_AS_ARRAY)
                                .readerFor(String[].class)
                                .with(CsvSchema.emptySchema().withSkipFirstDataRow(false));
try (
    InputStream in = fs.open(file);
    InputStream wrapped = wrap(in, configuration, file);
    MappingIterator<Object> iterator = headerReader.readValues(wrapped
) {
```

File: extensions/hadoop/src/main/java/com/hazelcast/jet/hadoop/impl/CsvInputFormat.java [hunk 2]
```diff
@@ -54,54 +59,87 @@ public RecordReader<NullWritable, Object> createRecordReader(InputSplit split, T
...
+                            return (String[]) iterator.next();
+                        }
+                    }
+                }
+
+            }
+
+            private InputStream wrap(InputStream in, Configuration configuration, Path file) throws IOException {
+                CompressionCodec codec = new CompressionCodecFactory(configuration).getCodec(file);
+                if (codec != null) {
```
Comment (TomaszGaweda) [inline extensions/hadoop/src/main/java/com/hazelcast/jet/hadoop/impl/CsvInputFormat.java:133]: nit: `return codec != null ? codec.createInputStream(in) : in`

File: extensions/hadoop/src/main/java/com/hazelcast/jet/hadoop/impl/CsvInputFormat.java [hunk 3]
```diff
@@ -110,19 +148,32 @@ public NullWritable getCurrentKey() {
             }
 
             @Override
-            public Object getCurrentValue() {
-                return current;
+            public Object getCurrentValue() throws IOException {
+                String current = reader.getCurrentValue().toString();
+                if (formatClazz.equals(String[].class)) {
+                    try (MappingIterator<Object> iterator = objectReader.readValues(current)) {
+                        return projection.apply(iterator.next());
```
Comment (TomaszGaweda) [inline extensions/hadoop/src/main/java/com/hazelcast/jet/hadoop/impl/CsvInputFormat.java:155]: no isNext guard?

File: extensions/hadoop/src/main/java/com/hazelcast/jet/hadoop/impl/CsvInputFormat.java [hunk 4]
```diff
@@ -110,19 +148,32 @@ public NullWritable getCurrentKey() {
             }
 
             @Override
-            public Object getCurrentValue() {
-                return current;
+            public Object getCurrentValue() throws IOException {
+                String current = reader.getCurrentValue().toString();
+                if (formatClazz.equals(String[].class)) {
```
Comment (TomaszGaweda) [inline extensions/hadoop/src/main/java/com/hazelcast/jet/hadoop/impl/CsvInputFormat.java:153]: nit: IIRC you can simply do formatClazz == String[].class

**Review by TomaszGaweda — APPROVED** [review]
Looks good, minor comments to consider

**Review by ldziedziul — APPROVED** [review]

File: extensions/hadoop/src/main/java/com/hazelcast/jet/hadoop/impl/CsvInputFormat.java (hunk 2 above)
Comment (frant-hartm) [reply extensions/hadoop/src/main/java/com/hazelcast/jet/hadoop/impl/CsvInputFormat.java]: Done in ec0def77

File: extensions/hadoop/src/main/java/com/hazelcast/jet/hadoop/impl/CsvInputFormat.java (hunk 3 above)
Comment (frant-hartm) [reply extensions/hadoop/src/main/java/com/hazelcast/jet/hadoop/impl/CsvInputFormat.java]: Done in ec0def77

File: extensions/hadoop/src/main/java/com/hazelcast/jet/hadoop/impl/CsvInputFormat.java (hunk 4 above)
Comment (frant-hartm) [reply extensions/hadoop/src/main/java/com/hazelcast/jet/hadoop/impl/CsvInputFormat.java]: Done in ec0def77

File: extensions/hadoop/src/test/java/com/hazelcast/jet/hadoop/file/CsvFileFormatTest.java [hunk 5]
```diff
@@ -129,4 +130,44 @@ public void shouldThrowWhenWrongFormatting() {
 
         assertJobFailed(source, CsvReadException.class, "Too many entries");
     }
+
+    @Test
+    public void shouldReadLargeCsvFile() throws Exception {
+        FileSourceBuilder<User> source = FileSources.files(currentDir + "/src/test/resources")
+                                                    .glob("file-large.csv")
```
[the remark this answers is not in the document]
Comment (frant-hartm) [reply extensions/hadoop/src/test/java/com/hazelcast/jet/hadoop/file/CsvFileFormatTest.java]: 1b14983b

File: extensions/hadoop/src/main/java/com/hazelcast/jet/hadoop/impl/CsvInputFormat.java (hunk 1 above)
Comment (frant-hartm) [reply extensions/hadoop/src/main/java/com/hazelcast/jet/hadoop/impl/CsvInputFormat.java]: Done in https://github.com/hazelcast/hazelcast/commit/ec0def77318fcb57eca1afa98af365acb984db8d

**Comment by frant-hartm** [comment]
Green build with http://jenkins.hazelcast.com/job/Hazelcast-EE-custom-tests-using-docker-jet-files-connector/6/

**Pull request merged**